File: spring-reactive-01/learn-java/src/main/java/boki/learnjava/context/api/ContextApiEx2.java
package boki.learnjava.context.api;

import boki.learnjava.util.Logger;
import boki.learnjava.util.TimeUtils;
import reactor.core.publisher.Mono;
import reactor.core.scheduler.Schedulers;
import reactor.util.context.Context;

/**
 * Context API 예제 코드
 * - pullAll(ContextView) API 사용
 */
public class ContextApiEx2 {
    public static void main(String[] args) {
        String key1 = "id";
        String key2 = "name";
        String key3 = "country";

        Mono.deferContextual(ctx ->
                Mono.just("ID: " + ctx.get(key1) + ", " + "Name: " + ctx.get(key2) +
                    ", " + "Country: " + ctx.get(key3))
            )
            .publishOn(Schedulers.parallel())
            .contextWrite(context -> context.putAll(Context.of(key2, "boki", key3, "Korea").readOnly()))
            .contextWrite(context -> context.put(key1, "개발꾼"))
            .subscribe(Logger::onNext);

        TimeUtils.sleep(100L);
    }
}

File: spring-reactive-01/learn-kt/src/main/kotlin/boki/learn/context/api/ContextApiEx2.kt
package boki.learn.context.api

import boki.learn.util.Logger
import reactor.core.publisher.Mono
import reactor.core.scheduler.Schedulers
import reactor.util.context.Context

/**
 * Context API 예제 코드
 * - pullAll(ContextView) API 사용
 */
fun main() {
    val key1 = "id"
    val key2 = "name"
    val key3 = "country"

    Mono.deferContextual { ctx ->
        val id: String = ctx[key1]
        val name: String = ctx[key2]
        val country: String = ctx[key3]

        Mono.just("ID: $id, Name: $name, Country: $country")
    }
        .publishOn(Schedulers.parallel())
        .contextWrite { it.putAll(Context.of(mapOf(key2 to "boki", key3 to "Korea")).readOnly()) }
        .contextWrite { it.put(key1, "개발꾼") }
        .subscribe { Logger.onNext(it) }

    Thread.sleep(100L)
}